README.md
# py_curl_example

To run this script type:

    python3 -m py_curl_example

It will generate a file named 'posts.csv'.


ChangeLog
CHANGES
=======

* Write basic howto
* Wite posts to CSV
* Parse all post entries
* Parsing page using BS4
* Download required URL using PyCurl
* Test bs4 library
* Initial commit
* Initial commit


py_curl_example/tests/test_bs.py
'''
Created on 15 Aug 2016

@author: fressi
'''

import os
import unittest

import bs4


HTML_DOC = """
<html><head><title>The Dormouse's story</title></head>
<body>
<p class="title"><b>The Dormouse's story</b></p>

<p class="story">Once upon a time there were three little sisters;
and their names were
<a href="http://example.com/elsie" class="sister" id="link1">Elsie</a>,
<a href="http://example.com/lacie" class="sister" id="link2">Lacie</a> and
<a href="http://example.com/tillie" class="sister" id="link3">Tillie</a>;
and they lived at the bottom of a well.</p>

<p class="story">...</p>
"""


TESTS_DIR = os.path.dirname(__file__)
REQUIRED_WEB_PAGE = os.path.join(TESTS_DIR, 'required_page.html')


class TestBeautifulSoup(unittest.TestCase):
    "Test case for bs4 library"

    def test_parse_page(self):
        "Test parsing a know web page."
        soup = bs4.BeautifulSoup(HTML_DOC, 'html.parser')

        self.assertEqual('title', soup.title.name)
        self.assertEqual("The Dormouse's story", soup.title.string)
        self.assertEqual('head', soup.title.parent.name)

    def test_parse_required_page(self):
        "Test parsing required web page."

        with open(REQUIRED_WEB_PAGE, encoding='iso-8859-1') as html_file:
            html = html_file.read()

        soup = bs4.BeautifulSoup(html, 'html.parser')

        self.assertEqual('title', soup.title.name)
        self.assertEqual("Classic Car Rescue - C5", soup.title.string)
        self.assertEqual('meta', soup.title.parent.name)


py_curl_example/tests/test_command.py
'''
Created on 15 Aug 2016

@author: fressi
'''
import os
import tempfile
import unittest

from py_curl_example import command


class TestCommand(unittest.TestCase):
    'Test case for command module'

    def test_get_html(self):
        'Test get_html function with required HTML and encoding'

        html = command.get_html(
            command.REQUIRED_URL, encoding=command.REQUIRED_ENCODING)

        self.assertTrue(html.endswith('</html>\n\n'), html[-30:])

    def test_get_page_entries(self):
        'Test get_html function with required HTML and encoding'

        entries = list(command.get_page_entries(
            command.REQUIRED_URL, encoding=command.REQUIRED_ENCODING))

        self.assertEqual(127, len(entries))

    def test_main(self):
        "Test main function"
        output_fd, output_path = tempfile.mkstemp()
        self.addCleanup(os.remove, output_path)
        os.close(output_fd)

        command.main(file_name=output_path)

        with open(output_path, 'rt', encoding='utf-8') as input_file:
            row0 = input_file.readline()
            self.assertTrue(
                row0.startswith(
                    '0;"Rick";"Mon Sep 24, 2012 4:53 pm";' +
                    '"Tonight, 8pm, might be worth a look...?\\r\\n\\n\\nRJ'),
                repr(row0))


py_curl_example/tests/test_py_curl.py
'''
Created on 15 Aug 2016

@author: fressi
'''
import unittest

import pycurl
import six


REQUIRED_URL = 'http://www.oldclassiccar.co.uk/forum/phpbb/phpBB2/' + \
               'viewtopic.php?t=12591'


class TestPyCurl(unittest.TestCase):
    "Test case for PyCurl library"

    def test_connect_some_web_site(self):
        "Test with connecting to pycurl with a know web page."
        buf = six.BytesIO()
        curl = pycurl.Curl()
        self.addCleanup(curl.close)
        curl.setopt(curl.URL, 'http://pycurl.io/')
        curl.setopt(curl.WRITEDATA, buf)

        curl.perform()
        body = buf.getvalue()

        self.assertTrue(body)

    def test_connect_required_web_site(self):
        "Test with connecting to pycurl with required web page."
        buf = six.BytesIO()
        curl = pycurl.Curl()
        self.addCleanup(curl.close)
        curl.setopt(curl.URL, REQUIRED_URL)
        curl.setopt(curl.WRITEDATA, buf)

        curl.perform()
        body = buf.getvalue()

        self.assertTrue(body)


py_curl_example/command.py
'''
Created on 15 Aug 2016

@author: fressi
'''

import csv
import logging
from urllib.parse import urljoin

import bs4
import pycurl
import six


LOG = logging.getLogger(__name__)

REQUIRED_URL = u'http://www.oldclassiccar.co.uk/forum/phpbb/phpBB2/' + \
               u'viewtopic.php?t=12591'
REQUIRED_ENCODING = 'iso-8859-1'


def get_html(url, encoding=None):
    "Gets an HTML from ginven url and translating it to a string."

    LOG.info('Get page from url: %s', url)

    html = six.BytesIO()
    curl = pycurl.Curl()
    try:
        curl.setopt(curl.URL, url)
        curl.setopt(curl.WRITEDATA, html)
        curl.perform()
        result = str(html.getvalue(), encoding=encoding)
        LOG.info('Page downloaded: %s', url)
        return result

    finally:
        curl.close()


def get_page_entries(url, encoding=None):
    "Parses given html."

    new_ulrs = {1: url}
    url_id = 1
    entry_id = 0

    while new_ulrs:
        url = new_ulrs.pop(url_id, None)
        url_id += 1

        html = get_html(url=url, encoding=encoding)
        soup = bs4.BeautifulSoup(html, 'html.parser')

        for entry in find_posted_entries(soup):
            yield (entry_id,) + entry
            entry_id += 1

        for new_url_id, new_url in find_page_links(soup):
            if new_url_id >= url_id and new_url_id not in new_ulrs:
                LOG.debug('New link found on page: %s -> %s', url, new_url)
                new_ulrs[new_url_id] = urljoin(url, new_url)


def find_page_links(soup):
    "Find all pages liks."
    for span in soup.find_all('span'):
        if span.get('class') == ['gensmall']:
            for a in span.find_all('a'):
                url = a.get('href')
                if url:
                    try:
                        url_id = int(a.get_text())
                    except ValueError:
                        continue
                    else:
                        yield url_id, url


def find_posted_entries(table):
    "Find all posted entries."
    for tr in table.find_all('tr'):
        fields = {}
        for td in tr.find_all('td'):
            for span in td.find_all('span'):
                cls = span.get('class')
                if cls:
                    fields[cls[0]] = span.get_text()

        name = fields.get('name')
        if name:
            post_details = fields.get('postdetails', '')
            if post_details:
                post_details = post_details.split('Posted:', 1)[1]
                post_details = post_details.strip()
                post_details = post_details.split('Post subject:', 1)[0]
                post_details = post_details.strip()
            post_body = repr(fields.get('postbody', ""))[1:-1]
            LOG.debug('%s: %s', name, post_details)
            yield name, post_details, post_body


def main(file_name='posts.csv'):
    "My script entry point."

    with open(file_name, 'wt', encoding='utf-8') as csv_file:
        csv_writer = csv.writer(
            csv_file, delimiter=';', escapechar='\\', doublequote=True,
            quoting=csv.QUOTE_NONNUMERIC)

        for entry in get_page_entries(url=REQUIRED_URL,
                                      encoding=REQUIRED_ENCODING):
            csv_writer.writerow(entry)
